src/Components/Server/src/Circuits/ServerComponentDeserializer.cs

```diff
@@ -25,8 +25,12 @@ namespace Microsoft.AspNetCore.Components.Server
     // 'sequence' indicates the order in which this component got rendered on the server.
     // 'assemblyName' the assembly name for the rendered component.
     // 'type' the full type name for the rendered component.
+    // 'parameterDefinitions' a JSON serialized array that contains the definitions for the parameters including their names and types and assemblies.
+    // 'parameterValues' a JSON serialized array containing the parameter values.
     // 'invocationId' a random string that matches all components rendered by as part of a single HTTP response.
     // For example: base64(dataprotection({ "sequence": 1, "assemblyName": "Microsoft.AspNetCore.Components", "type":"Microsoft.AspNetCore.Components.Routing.Router", "invocationId": "<<guid>>"}))
+    // With parameters
+    // For example: base64(dataprotection({ "sequence": 1, "assemblyName": "Microsoft.AspNetCore.Components", "type":"Microsoft.AspNetCore.Components.Routing.Router", "invocationId": "<<guid>>", parameterDefinitions: "[{ \"name\":\"Parameter\", \"typeName\":\"string\", \"assembly\":\"System.Private.CoreLib\"}], parameterValues: [<<string-value>>]}))
```

Comment:
How often do we expect to run into the message size limit - what we have tested this with?

Reply:
I'm not advocating to do something more complicated unless we collect data that motivates it - I want to understand what the limits are for now because we have to explain to people how to use this feature.

Reply:
Since the default message size is 32KB, you wouldn't really be able to come anywhere close if all you're passing is something like IDs to entities in a database, or a bunch of flags or filename strings.

The most natural case where you could hit it is if you're performing something like a DB query during `_Host.cshtml` render-time, and trying to pass the complete query results into your component. Maybe someone does this because they're trying to run the query only once and have it produce prerendered output and then the same query results become part of the interactive component without re-querying.

I have a bit of sympathy for that scenario, but still don't think it warrants doing something more complicated here in 3.1 without really clear customer feedback and validation of these scenarios. Developers in this scenario should consider doing the query results caching at a different level, e.g., in some kind of repository.

Reply:
Perhaps one other scenario is super-fine-grained components. Consider a server-rendered grid, in which every cell is a separate interactive component with its own set of parameters. If you have thousands of such cells, that probably exceeds the message size limit.

My recommendation in this case would be not to be so super-fine-grained at the root level, and instead just have a single top-level "grid" component (which in turn can render as many thousands of child components as it wants).

It would be nice to know if the failure modes are understandable. My guess is that they won't be. I don't know what SignalR's error message for "incoming message too big" looks like, but it's probably hard to tie that back to the developer's choice of root-component granularity.

Reply:
> Maybe someone does this because they're trying to run the query only once and have it produce prerendered output and then the same query results become part of the interactive component without re-querying.

I actually expect a lot of users to do that. In my experience if you give users a magic portal to shove objects into they will shove lots of stuff through. I want to know that we've actually thought through scenarios and tried this before we merge it.

Reply:
I'm going to try and summarize my thoughts on this a bit.
* With regards to limits (size) there's always someone who is going to hit them.
* I don't expect most users hitting it, based on the fact that is quite high, although definitely possible.
* For users that do, its important that we understand the scenarios and have alternative guidance on how to achieve what they want and leave that clear in the docs.
* The size limitation is a restriction of the current implementation that can be further lifted in the future in a non breaking way.

With regards to scenarios, I believe there are two main scenarios.
* You are rendering a server-side blazor app and want to pass in initial configuration/parameters to your app.
  * I would estimate configuration data to be at most a few KBs (4, 8) is my current estimation.
  * If you hit this limit, you can always store the config in memory and serialize a "reference/pointer/id" to it that is used to retrieve it.
* You have an MVC/Razor page with components "sprinkled" along the page.
  * For people running into this we would give them the same guidance @SteveSandersonMS suggested.

I'll make sure all these things are covered in docs. I've filed an issue https://github.com/aspnet/AspNetCore.Docs/issues/14878 to track it.

The key thing to understand is that given that the information gets serialized you can always serialize that information beforehand and change your component to receive a handle to the data that it uses to retrieve and render it.